---
title: "RNA_SEQ_LAB3"
output: html_notebook
---

```{r}
library(dplyr)
library(Seurat)
library(patchwork)

# Load the PBMC dataset
pbmc.data <- Read10X(data.dir = "/Users/kaylapoirier/Desktop/filtered_gene_bc_matrices/hg19") #read10x reads output which returns a unique moecular identified count matrix (UMI)
# Initialize the Seurat object with the raw (non-normalized data).
pbmc <- CreateSeuratObject(counts = pbmc.data, project = "pbmc3k", min.cells = 3, min.features = 200)
pbmc
```

# Standard pre-processing workflow
preprocessing for scRNA-seq data in Seurat

## QC and selecting cells for further analysis

```{r}
# The [[ operator can add columns to object metadata. This is a great place to stash QC stats
pbmc[["percent.mt"]] <- PercentageFeatureSet(pbmc, pattern = "^MT-") #calculates percentage of counts origninaing fro a set of features
```

```{r}
# Visualize QC metrics as a violin plot
VlnPlot(pbmc, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3) #
```

```{r}
# FeatureScatter is typically used to visualize feature-feature relationships, but can be used
# for anything calculated by the object, i.e. columns in object metadata, PC scores etc.

plot1 <- FeatureScatter(pbmc, feature1 = "nCount_RNA", feature2 = "percent.mt") #comparing percent mt with ncount RNA
plot2 <- FeatureScatter(pbmc, feature1 = "nCount_RNA", feature2 = "nFeature_RNA") #comparing n feature to n count RNA
plot1 + plot2

pbmc <- subset(pbmc, subset = nFeature_RNA > 200 & nFeature_RNA < 2500 & percent.mt < 5)
```

# Normalizing the Data

Use global-scaling normalization method. Assumes that each cell has the same number of RNA molecules

```{r}
pbmc <- NormalizeData(pbmc, normalization.method = "LogNormalize", scale.factor = 10000) #NormalizeData funciton that normalizes feature expression measurements for each cell by the total expression and multiple by a scale of 10000. it thrn log transform

pbmc <- NormalizeData(pbmc) #without parameters
```

# Identification of highly variable features (feature selection)

Calculate subset of features with high variation in different cells

```{r}
pbmc <- FindVariableFeatures(pbmc, selection.method = "vst", nfeatures = 2000) #return 2000 features that calculates mean variance 

# Identify the 10 most highly variable genes
top10 <- head(VariableFeatures(pbmc), 10)

# plot variable features with and without labels
plot1 <- VariableFeaturePlot(pbmc)
plot2 <- LabelPoints(plot = plot1, points = top10, repel = TRUE)
plot1 + plot2
```

# Scaling the data

```{r}
all.genes <- rownames(pbmc)
pbmc <- ScaleData(pbmc, features = all.genes) #use ScaleData that did a linear transformation. This provides an ewual weight to all genes 
```

# Perform linear dimensional reduction

Perform PCA

```{r}
pbmc <- RunPCA(pbmc, features = VariableFeatures(object = pbmc)) #using variable features 

# Examine and visualize PCA results a few different ways
print(pbmc[["pca"]], dims = 1:5, nfeatures = 5) # can also use VizDimReduction(), DimPlot(), and DimHeatmap()
```

```{r}
VizDimLoadings(pbmc, dims = 1:2, reduction = "pca")
```

```{r}
DimPlot(pbmc, reduction = "pca") + NoLegend()
```

```{r}
DimHeatmap(pbmc, dims = 1, cells = 500, balanced = TRUE) #visualize heterogeneity which can be used to decide what PC is best 
```

```{r}
DimHeatmap(pbmc, dims = 1:15, cells = 500, balanced = TRUE)
```

# Determine the ‘dimensionality’ of the dataset

Metafeatures are PCs and clusters cells by PCA scores which then cmpresses removes the noise from the dataset

```{r}
ElbowPlot(pbmc) #ranking PCs based on variance. Based off this data, the elbow is at 5 and so a majority of true signal is captured in the first 5 PCs
```

# Cluster the cells
first construct a K-nearest neighbor (KNN) graph using Euclidean distance 
```{r}
pbmc <- FindNeighbors(pbmc, dims = 1:5) #takes first 5 PCs 
pbmc <- FindClusters(pbmc, resolution = 0.5) #group cells together
```

```{r}
# Look at cluster IDs of the first 5 cells
head(Idents(pbmc), 5) #use Indents to visualize it
```
# Run non-linear dimensional reduction (UMAP/tSNE)

nonlinear domenstional to techniques to visualize data
```{r}
pbmc <- RunUMAP(pbmc, dims = 1:5)
```

```{r}
# note that you can set `label = TRUE` or use the LabelClusters function to help label
# individual clusters
DimPlot(pbmc, reduction = "umap")
```

```{r}
saveRDS(pbmc, file = "pbmc_tutorial.rds") #save plot
```

# Finding differentially expressed features (cluster biomarkers)

```{r}
# find all markers of cluster 2
cluster2.markers <- FindMarkers(pbmc, ident.1 = 2) #identifies positive and negative markers of a single cluster compared to other cells and gives p values, log2FC for each gene
head(cluster2.markers, n = 5)
```

```{r}
# find all markers distinguishing cluster 5 from clusters 0 and 3
cluster5.markers <- FindMarkers(pbmc, ident.1 = 5, ident.2 = c(0, 3))
head(cluster5.markers, n = 5)
```

```{r}
# find markers for every cluster compared to all remaining cells, report only the positive
# ones
pbmc.markers <- FindAllMarkers(pbmc, only.pos = TRUE)
pbmc.markers %>%
    group_by(cluster) %>%
    dplyr::filter(avg_log2FC > 1)
```

```{r}
cluster0.markers <- FindMarkers(pbmc, ident.1 = 0, logfc.threshold = 0.25, test.use = "roc", only.pos = TRUE) #use test.use to set tests for differential expression
```

```{r}
VlnPlot(pbmc, features = c("MS4A1", "CD79A")) #shows expression probability distributions across clusters of MS4A1 and CD79A genes
```

```{r}
# you can plot raw counts as well
VlnPlot(pbmc, features = c("NKG7", "PF4"), slot = "counts", log = TRUE)
```

```{r}
FeaturePlot(pbmc, features = c("MS4A1", "GNLY", "CD3E", "CD14", "FCER1A", "FCGR3A", "LYZ", "PPBP",
    "CD8A")) #visualizes feature expression on a tSNE or PCA plot
```

```{r}
pbmc.markers %>%
    group_by(cluster) %>%
    dplyr::filter(avg_log2FC > 1) %>%
    slice_head(n = 10) %>%
    ungroup() -> top10
DoHeatmap(pbmc, features = top10$gene) + NoLegend() #heatmaps for given cells and features of 20 markers 
```

# Assigning cell type identity to clusters

```{r}
new.cluster.ids <- c("Naive CD4 T", "CD14+ Mono", "Memory CD4 T", "B", "CD8 T", "FCGR3A+ Mono",
    "NK", "DC", "Platelet")
names(new.cluster.ids) <- levels(pbmc)
pbmc <- RenameIdents(pbmc, new.cluster.ids)
DimPlot(pbmc, reduction = "umap", label = TRUE, pt.size = 0.5) + NoLegend() #use markers to match clustering
```

```{r}
library(ggplot2)
plot <- DimPlot(pbmc, reduction = "umap", label = TRUE, label.size = 4.5) + xlab("UMAP 1") + ylab("UMAP 2") +
    theme(axis.title = element_text(size = 18), legend.text = element_text(size = 18)) + guides(colour = guide_legend(override.aes = list(size = 10)))
ggsave(filename = "../output/images/pbmc3k_umap.jpg", height = 7, width = 12, plot = plot, quality = 50)

saveRDS(pbmc, file = "../output/pbmc3k_final.rds") #save 
```