# Bioinformatics-Code

Poirier_04A_Biopython_Tutorial_LSH.ipynb is a Biopython tutorial that includes reading fasta and genebank files, working with feature tables and producing phylogenetic trees. Files required to run Poirier_04A_Biopython_Tutorial_LSH.ipynb are KRAS.fasta, KRAS_genbank.gb and opuntia.dnd. 

Poirier_QiimeTutorialNotes.tex is a text file that follows a Qiime2 tutorial from https://docs.qiime2.org/2024.10/tutorials/moving-pictures/. It provides information on the basic steps on how to use Qiime such as cleaning data, generating feature tables, making phylogenetic analyses, alpha and beta diversity analyses and making emperor plots. 

Poirier_Seurat.Rmd is a R notebook that runs through a tutorial on Single-Cell RNA sequence analysis using Seurat from Seurat-Guided tutorial -2,700 PBMCs. The tutorial goes through preprocessing workflow in Seurat, normalizing data, identifying variable features, scaling data, performing PCA analysis, clustering cells and running UMAO and tSNE. pbmc3k_filtered_gene_bc_matrices.tar.gz is required to run this notebook and needs to be uncompressed. 

Poirier_DESEQ2.RMd is a R notebook that runs through a tutorial using DESeq2 which is used for RNA-seq and differential gene expression. This code uses the dataset from files rna_counts_data.csv and rna_map_updata_copy.csv to gain an understanding of differences between mutant cancer cells and normal cells from the same line. The tutorial is by Ashley Valentina Schwartz, PhD from https://ashleyschwartz.com/posts/2023/05/deseq2-tutorial. 

Poirier_LimmaGlimmaEdge.Rmd is a R notebook that runs through a tutorial that uses limma, Glimma and edgeR to do an RNA-sequencing analysis on cell populations from the mammary glands of mice. The tutorial is from Law et al (2018) https://bioconductor.org/packages/release/workflows/vignettes/RNAseq123/inst/doc/limmaWorkflow.html.


